vision/deskew/vector_utils.py

```diff
@@ -0,0 +1,274 @@
+"""Functions that use vectors to calculate camera intersections with the ground"""
+
+from typing import TypeAlias
+from nptyping import NDArray, Shape, Float64
+import numpy as np
+from scipy.spatial.transform import Rotation
+
+Vector: TypeAlias = NDArray[Shape["3"], Float64]
+from vision.constants import Point, SENSOR_WIDTH, SENSOR_HEIGHT, ROTATION_OFFSET
+
+
+# Vector pointing toward the +X axis, represents the camera's forward direction when the rotation
+#   on all axes is 0
+IHAT: Vector = np.array([1, 0, 0], dtype=np.float64)
+
+
+def pixel_intersect(
+    pixel: tuple[int, int],
+    image_shape: tuple[int, int, int] | tuple[int, int],
+    focal_length: float,
+    rotation_deg: list[float],
+    height: float,
+) -> Point | None:
+    """
+    Finds the intersection [X,Y] of a given pixel with the ground relative to the camera.
+    A camera with no rotation points in the +X direction and is centered at [0, 0, height].
+
+    Parameters
+    ----------
+    pixel : tuple[int, int]
+        The coordinates of the pixel in [Y, X] form
+    image_shape : tuple[int, int, int] | tuple[int, int]
+        The shape of the image (returned by image.shape when image is a numpy image array)
+    focal_length : float
+        The camera's focal length
+    rotation_deg : list[float]
+        The [roll, pitch, yaw] rotation in degrees
+    height : float
+        The height that the image was taken at. The units of the output will be the units of the
+        input.
+
+    Returns
+    -------
+    intersect : Point | None
+        The coordinates [X,Y] where the pixel's vector intersects with the ground.
+        Returns None if there is no intersect.
+    """
+
+    # Create the normalized vector representing the direction of the given pixel
+    vector: Vector = pixel_vector(pixel, image_shape, focal_length)
+
+    rotation_deg = np.deg2rad(rotation_deg).tolist()
+
+    vector = euler_rotate(vector, rotation_deg)
+
+    # Apply the constant rotation offset
+    vector = euler_rotate(vector, ROTATION_OFFSET)
+
+    intersect: Point | None = plane_collision(vector, height)
+
+    return intersect
+
+
+def plane_collision(ray_direction: Vector, height: float) -> Point | None:
+    """
+    Returns the point where a ray intersects the XY plane
+    Returns None if there is no intersect.
+
+    Parameters
+    ----------
+    ray_direction : Vector
+        XYZ coordinates that represent the direction a ray faces from (0, 0, 0)
+    height : float
+        The Z coordinate for the starting height of the ray; can be any units
+
+    Returns
+    -------
+    intersect : Point | None
+        The ray's intersection with the plane in [X,Y] format
+        Returns None if there is no intersect.
+
+    """
+    # Find the "time" at which the line intersects the plane
+    # Line is defined as ray_direction * time + origin.
+    # Origin is the point at X, Y, Z = (0, 0, height)
+
+    intersect: Point | None
+
+    intersect = None
```

Comment:
You have this assignment right after you declare its types. Should combine into one statement.